### ISSUE TITLE
Multiple worker support

### ISSUE THREAD

**Issue by richardrl** [vitchyr/rlkit#27]
Is there support for multiple workers (threads each feeding in independent samples)?

I'm trying to reproduce the pick and place here, and they mention needing 19 workers to do so: https://github.com/openai/baselines/tree/master/baselines/her

On the other hand, seems like since HER is off policy multiple workers aren't necessary..

**Comment by vitchyr** [comment]
No. However, if you're interested in implementing it, I would start with the batch training mode and just replace the "_take_step_in_env" call with some parallelized implementation.

**Comment by vitchyr** [comment]
I'm not planning on adding this any time soon, though I've been talking with some people at Berkeley who might be interested in doing this, either as part of this repo or separately. PRs would also be welcomed, and if you're interested in this let me know.

**Comment by vitchyr** [comment]
I'm going to re-open this and label it as a enhancement request. Though like I said, I probably won't get around to it any time soon.

**Comment by bycn** [comment]
Is there any progress on this?

**Comment by richardrl** [comment]
@bycn You can check how I implemented it here: https://github.com/richardrl/rlkit-relational/blob/master/rlkit/torch/optim/mpi_adam.py

**Comment by bycn** [comment]
Thanks for the reply—actually, I realized I'm looking for parallelized sampling, not parallelized consumption.

**Comment by richardrl** [comment]
@bycn Parallel sampling is also implemented. Check how "num_parallel_processes" is used here: https://github.com/richardrl/rlkit-relational/blob/master/examples/relationalrl/train_pickandplace1.py

**Comment by bycn** [comment]
It's unclear to me how parallel sampling is implemented here without vectorized environments, i.e., like https://stable-baselines.readthedocs.io/en/master/guide/vec_envs.html. Can you explain the approach?

**Comment by richardrl** [comment]
I don't know what the vectorized environments are doing, but I am simply starting a Mujoco instance in multiple threads with a local replay buffer in each thread, sampling each environment to fill the local replay buffer, computing the gradient, and then taking the average of the gradient in the optimizer file you see above.

**Comment by richardrl** [comment]
If you have further questions about my implementation, feel free to open an issue in that repo.
